### DISCUSSION TITLE
ValueWrapper

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#216] [General]
I remember talking with you about something related.

Having `ValueWrapper` as `typedef ValueWrapper<T> = (T Function(), void Function(T));`.

I came to situations where I need to know the result of the setter function, whether the value changed or it's considered to be the same.

This would require a minimal change to: `typedef ValueWrapper<T> = (T Function(), bool Function(T));`.

It would help to avoid boilerplate (wrappers around that function).

**Comment by GregoryConrad** [comment]
Not sure I’d want to make this change. You normally shouldn’t need to check the return value of a setter in ReArch since that often leads to more imperative code. can you share a use case?

anyways, this is easy enough to do via a new side effect and a new type def defined in your code. Just not sure I want to commit to it in ReArch.

**Reply by busslina** [reply]
This is in my `App` `RearchComponent`:

```dart
// Router delegate listener
    use.effect(
      () {
        void listener() {
          final currentPath = routerDelegate!.currentUri.toString();

          log.debug('Router path: $currentPath', active: debugInfo.routerPath);

          // Caching current path
          if (!use(currentPathCapsule).$2(currentPath)) return;

          // Sending current path via websocket
          websocket.dataUpdate(
            dataCode: DataCodes.clientLiveMonitorization,
            encodedData: ClientMonitorizationUpdateRequest(
              currentUri: currentPath,
            ).serialize(),
          );
        }

        if (routerDelegate != null) {
          routerDelegate.addListener(listener);
          return () => routerDelegate.removeListener(listener);
        }

        return null;
      },
      [routerDelegate],
    );
```

The beamer delegate is executing its listeners on events that not necessarily implies a change on the route path.

So knowing whether the setter successfully change its value helps me to avoid informing the backend about a duplicated route change.

For context about this use.effect: I dispose/regenerate the Router Delegate when websocket connection is lost.

I understand your POV about this, but I'm talking about making easy to engage declarative (ReArch) with some other approaches that are more imperative.

**Reply by busslina** [reply]
> anyways, this is easy enough to do via a new side effect and a new type def defined in your code. Just not sure I want to commit to it in ReArch.

Yes, but as I use a lot `ValueWrapper` I have a lot of extension utils on it, forcing me to duplicate each one.

Anyway, I undestand your POV.

**Reply by busslina** [reply]
And yes, I'm seeing now that I could just inform the backend on a capsule that listen `currentPathCapsule` (Or inside `currentPathCapsule`).

**Reply by GregoryConrad** [reply]
> Yes, but as I use a lot ValueWrapper I have a lot of extension utils on it, forcing me to duplicate each one.

you’d have to try, but I believe extensions on ValueWrapper would also work for a new type def returning a bool for second function. That’s because a bool Function() can always be used where a void Function() is expected
